rclpy/src/rclpy/_rclpy_action.c

```diff
@@ -0,0 +1,929 @@
+  }
+
+  if (RCL_RET_OK != ret) {
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to get number of entities for '%s': %s",
+      PyCapsule_GetName(pyentity),
+      rcl_get_error_string().str);
+    rcl_reset_error();
+    return NULL;
+  }
+
+  PyObject * result_tuple = PyTuple_Pack(5,
+      PyLong_FromSize_t(num_subscriptions),
+      PyLong_FromSize_t(num_guard_conditions),
+      PyLong_FromSize_t(num_timers),
+      PyLong_FromSize_t(num_clients),
+      PyLong_FromSize_t(num_services));
+  if (!result_tuple) {
+    return PyErr_Format(PyExc_RuntimeError, "Failed to create new tuple object");
+  }
+  return result_tuple;
+}
+
+/// Check if an action entity has any ready wait set entities.
+/**
+ * This must be called after waiting on the wait set.
+ * Raises RuntimeError on failure.
+ *
+ * \param[in] entity Capsule pointing to the action entity
+ *   (rcl_action_client_t or rcl_action_server_t).
+ * \param[in] pywait_set Capsule pointing to the wait set structure.
+ * \return A tuple of Bool representing the ready sub-entities.
+ *     For a rcl_action_client_t:
+ *       (is_feedback_ready,
+ *        is_status_ready,
+ *        is_goal_response_ready,
+ *        is_cancel_response_ready,
+ *        is_result_response_ready)
+ *
+ *     For a rcl_action_server_t:
+ *       (is_goal_request_ready,
+ *        is_cancel_request_ready,
+ *        is_result_request_ready,
+ *        is_goal_expired)
+ */
+static PyObject *
+rclpy_action_wait_set_is_ready(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pyentity;
+  PyObject * pywait_set;
+
+  if (!PyArg_ParseTuple(args, "OO", &pyentity, &pywait_set)) {
+    return NULL;
+  }
+
+  rcl_wait_set_t * wait_set = (rcl_wait_set_t *)PyCapsule_GetPointer(pywait_set, "rcl_wait_set_t");
+  if (!wait_set) {
+    return NULL;
+  }
+
+  if (PyCapsule_IsValid(pyentity, "rcl_action_client_t")) {
+    rcl_action_client_t * action_client =
+      (rcl_action_client_t *)PyCapsule_GetPointer(pyentity, "rcl_action_client_t");
+    bool is_feedback_ready = false;
+    bool is_status_ready = false;
+    bool is_goal_response_ready = false;
+    bool is_cancel_response_ready = false;
+    bool is_result_response_ready = false;
+    rcl_ret_t ret = rcl_action_client_wait_set_get_entities_ready(
+      wait_set,
+      action_client,
+      &is_feedback_ready,
+      &is_status_ready,
+      &is_goal_response_ready,
+      &is_cancel_response_ready,
+      &is_result_response_ready);
+    if (RCL_RET_OK != ret) {
+      PyErr_Format(PyExc_RuntimeError,
+        "Failed to get number of ready entities for action client: %s",
+        rcl_get_error_string().str);
+      rcl_reset_error();
+      return NULL;
+    }
+
+    PyObject * result_tuple = PyTuple_Pack(5,
+        PyBool_FromLong(is_feedback_ready),
+        PyBool_FromLong(is_status_ready),
+        PyBool_FromLong(is_goal_response_ready),
+        PyBool_FromLong(is_cancel_response_ready),
+        PyBool_FromLong(is_result_response_ready));
```

Comment:
Memory is leaked if `PyBool_FromLong` succeeds but `PyTuple_Pack` fails

Reply:
Fixed 34db026